src/variables/components/TypeAheadVariableDropdown.tsx

```diff
@@ -0,0 +1,308 @@
+    const {actualVal, loaded, selectHappened, menuOpen} = this.state
+    const {
+      actualVal: prevActualVal,
+      selectHappened: prevSelectHappened,
+      menuOpen: prevMenuOpen,
+    } = prevState
+
+    // this is for updating the values:
+    // (only want this to run *once* when the values get loaded)
+    if (!loaded && prevVals.length !== values.length) {
+      this.setState({
+        shownValues: values,
+        typedValue: selectedValue,
+        loaded: true,
+      })
+    }
+
+    // unset the menuOpen; it should be set to closed (or open) only once; then undone
+    if (menuOpen !== prevMenuOpen && menuOpen !== null) {
+      this.setState({menuOpen: null})
+    }
+
+    // need to have this, as the 'onClickAwayHere' gets triggered *before*
+    // the selected value is set to the actualValue (it keeps re-using the original
+    // property)
+
+    // for updating the selected value:
+    if (
+      selectHappened &&
+      !prevSelectHappened &&
+      actualVal &&
+      actualVal !== prevActualVal
+    ) {
+      this.setState({typedValue: actualVal, selectHappened: false})
+    }
+  }
+
+  filterVals = needle => {
+    const {values} = this.props
+
+    if (!needle) {
+      this.setState({shownValues: values, typedValue: needle})
+    } else {
+      const result = values.filter(val =>
+        val.toLowerCase().includes(needle.toLowerCase())
+      )
+      this.setState({
+        shownValues: result,
+        typedValue: needle,
+        menuOpen: MenuStatus.Open,
+      })
+    }
+  }
+
+  maybeSelectNextItem = e => {
+    const {shownValues, selectIndex} = this.state
+
+    let newIndex = -1
+
+    if (e.keyCode === 40) {
+      // down arrow
+      newIndex = selectIndex + 1
+    } else if (e.keyCode === 38) {
+      // up arrow
+      newIndex = selectIndex - 1
+    }
+
+    const numItems = shownValues.length
+
+    if (numItems && newIndex >= 0 && newIndex < numItems) {
+      this.handleSelect(shownValues[newIndex], newIndex)
+      return
+    }
+
+    if (e.keyCode === 13) {
+      // return/enter key
+      // lose focus, reset the selectIndex to -1, & close the menu:
+      e.target.blur()
+
+      // the person could have been typing and pressed return, need to reset the value
+      // back to the 'real value'
+      const newState = {
+        menuOpen: MenuStatus.Closed,
+        selectIndex: -1,
+        ...this.getRealValue(),
+      }
+      this.setState(newState)
+    }
+  }
+
+  // if the entire item loses focus, then
+  // the input should show the actual selected item, not what the user typed in;
+  // only want to show valid values when the component is not actively being used
+  onClickAwayHere = () => {
+    this.setState(this.getRealValue())
+  }
+
+  getRealValue = () => {
+    const {actualVal, typedValue} = this.state
+    const {selectedValue} = this.props
+
+    if (actualVal || selectedValue) {
+      const realValue = actualVal ? actualVal : selectedValue
```

Comment:
you could shorten this to `const realValue = actualVal ?? selectedValue` if you wanted

Reply:
done.